### PR TITLE
[HER-48] Create detection seeds

### PR DESCRIPTION
**Pull request by TBiannca**
## Description

- Create seeds infra
- Create detections seed
- It's gonna replace the previously created detection migration

## Related
[HER-48](https://nerds-interns.atlassian.net/browse/HER-48?atlOrigin=eyJpIjoiYWY2ZWFlZjNjNTgzNGQ3N2ExN2I0OTczYzE3MDNlZDEiLCJwIjoiaiJ9)

### CODE REVIEW

File: src/Business/Seeds/DetectionSeed.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                _repository.Insert(missingDetection);
+            }
+        }
+
+        private IEnumerable<MCreateDetection> MissingDetections()
+        {
+            var detections = _dataFactory.Make<MCreateDetection>("detections.json")
+                .Where(x => !_repository.ByInput(x).Any());
+
+            return detections;
```
Comment (RaduBuciuceanu) [inline src/Business/Seeds/DetectionSeed.cs:33]: Should we remove this useless variable here?

File: src/Business/Seeds/DetectionSeed.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            foreach (var missingDetection in missingDetections)
+            {
+                _repository.Insert(missingDetection);
+            }
+        }
+
+        private IEnumerable<MCreateDetection> MissingDetections()
+        {
+            var detections = _dataFactory.Make<MCreateDetection>("detections.json")
+                .Where(x => !_repository.ByInput(x).Any());
```
Comment (RaduBuciuceanu) [inline src/Business/Seeds/DetectionSeed.cs:31]: Should we try finding a more suggestive name for this magic `x`?

File: src/Business/Seeds/ISeed.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+namespace Business.Seeds
+{
+public interface ISeed
```
Comment (RaduBuciuceanu) [inline src/Business/Seeds/ISeed.cs:3]: Should we fix the indentation here?

File: src/Data/Detection/DetectionRepository.cs [hunk 4]
```diff
@@ -26,5 +26,20 @@ public IEnumerable<MDetection> ByParameter() => _context.Set<EDetection>()
...
+            return DetectionFactory.MakeModel(result.Entity);
+        }
+
+        public IEnumerable<MDetection> ByInput(MCreateDetection detection)
+        {
+            return _context.Set<EDetection>()
+                .Where(entity => detection.Class == entity.Class && detection.Score == entity.Score)
+                .ToList()
+                .Select(DetectionFactory.MakeModel);
+        }
```
Comment (RaduBuciuceanu) [inline src/Data/Detection/DetectionRepository.cs:43]: Should we convert this method into a one-liner?

File: src/Seeds/Data/detections.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  "class": "boat",
+  "score": 1,
+  "timestamp": "5/1/2021 8:30:52AM"
+},
+{
+"class": "glass",
+"score": 7,
+"timestamp": "5/1/2021 8:30:52AM"
+}]
+
```
Comment (RaduBuciuceanu) [inline src/Seeds/Data/detections.json:11]: Should we consider adding more items here? Maybe about 50 of them?

**Review by RaduBuciuceanu — CHANGES_REQUESTED** [review]

File: src/Presentation/Startup.cs [hunk 6]
```diff
@@ -29,13 +31,19 @@ public void ConfigureServices(IServiceCollection services)
...
-        public void Configure(IApplicationBuilder app, Context context)
+        public void Configure(IApplicationBuilder app, Context context, IEnumerable<ISeed> seeds)
         {
             ConfigureEnvironment(app);
             ConfigureDatabase(context);
             ConfigureEndpoints(app);
+            foreach (var seed in seeds)
+            {
+                seed.Execute();
+            }
```
[the remark this answers is not in the document]
Comment (TBiannca) [reply src/Presentation/Startup.cs:48]: done

File: src/Seeds/Seeds.csproj [hunk 7]
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net5.0</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\Business\Business.csproj" />
+      <ProjectReference Include="..\Data\Data.csproj" />
```
[the remark this answers is not in the document]
Comment (TBiannca) [reply src/Seeds/Seeds.csproj:9]: done

File: src/Business/Seeds/DetectionSeed.cs (hunk 1 above)
Comment (TBiannca) [reply src/Business/Seeds/DetectionSeed.cs:33]: done

File: src/Business/Seeds/DetectionSeed.cs (hunk 2 above)
Comment (TBiannca) [reply src/Business/Seeds/DetectionSeed.cs:31]: done

File: src/Business/Seeds/ISeed.cs (hunk 3 above)
Comment (TBiannca) [reply src/Business/Seeds/ISeed.cs:3]: done

File: src/Data/Detection/DetectionRepository.cs (hunk 4 above)
Comment (TBiannca) [reply src/Data/Detection/DetectionRepository.cs:43]: done

File: src/Seeds/Data/detections.json (hunk 5 above)
Comment (TBiannca) [reply src/Seeds/Data/detections.json:261]: done

File: src/Data/Detection/DetectionRepository.cs [hunk 8]
```diff
@@ -16,16 +17,26 @@ public IEnumerable<MDetection> ByParameter() => _context.Set<EDetection>()
...
-            .Select(DetectionFactory.MakeModel);
+        public IEnumerable<MDetection> Insert(IEnumerable<MCreateDetection> input) => new List<MDetection>
+        {
+            new MDetection
+            {
+                Class = "glass",
+                Score = 100,
+                Timestamp = DateTime.UtcNow,
+            },
+        };
```
Comment (RaduBuciuceanu) [inline src/Data/Detection/DetectionRepository.cs:28]: This one will overwrite @ConstantinescuAndrei work. Should we rollback it?

File: src/Seeds/DataFactory.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    using System.Collections.Generic;
+    using System.IO;
+    using Business.Seeds;
+    using Newtonsoft.Json;
+
+    public class DataFactory : IDataFactory
+    {
+        public IEnumerable<TModel> Make<TModel>(string path)
+        {
+            string fullPath = $"../Seeds/Data/{path}";
```
Comment (RaduBuciuceanu) [inline src/Seeds/DataFactory.cs:12]: This one fails the deployment. Let's find out a fix for it tomorrow.

**Review by RaduBuciuceanu — CHANGES_REQUESTED** [review]

File: src/Business/Detection/Common/Repositories/IDetectionRepository.cs [hunk 10]
```diff
@@ -9,5 +9,9 @@ public interface IDetectionRepository
         public IEnumerable<MDetection> ByParameter();
 
         public IEnumerable<MDetection> Insert(IEnumerable<MCreateDetection> input);
+
+        public MDetection Insert(MCreateDetection input);
```
Comment (danidoff) [inline src/Business/Detection/Common/Repositories/IDetectionRepository.cs:13]: Should we delete this method and use the other Insert which will insert a list of inputs?

Comment (danidoff) [inline src/Business/Detection/Common/Repositories/IDetectionRepository.cs:13]: Should we delete this method and use the other Insert which will insert a list of inputs?

File: src/Business/Seeds/DetectionSeed.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        public DetectionSeed(IDetectionRepository repository, IDataFactory dataFactory)
+        {
+            _repository = repository;
+            _dataFactory = dataFactory;
+        }
+
+        public void Execute()
+        {
+            var missingDetections = MissingDetections();
+            foreach (var missingDetection in missingDetections)
```
Comment (danidoff) [inline src/Business/Seeds/DetectionSeed.cs:22]: Should we delete this foreach and use directly `_repository.Insert(missingDetections)` ?

File: src/Business/Seeds/DetectionSeed.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+            {
+                _repository.Insert(missingDetection);
+            }
+        }
+
+        private IEnumerable<MCreateDetection> MissingDetections()
```
Comment (danidoff) [inline src/Business/Seeds/DetectionSeed.cs:28]: Should we rename the `MCreateDetection` model into `MCreateDetections`? Our Insert from repo will insert a list of detections not a single one and that`s why we need to rename the model.

**Review by danidoff — CHANGES_REQUESTED** [review]

File: src/Business/Seeds/DetectionSeed.cs (hunk 12 above)
Comment (TBiannca) [reply src/Business/Seeds/DetectionSeed.cs:28]: done

**Review by RaduBuciuceanu — APPROVED** [review]

**Review by Urzica-Platon-Casian — DISMISSED** [review]
Looks good!

**Review by RaduBuciuceanu — APPROVED** [review]

**Review by danidoff — APPROVED** [review]

**Pull request merged**